src/v/redpanda/admin_server.h

```diff
@@ -147,27 +148,32 @@ class admin_server {
           [this, handler](std::unique_ptr<ss::http::request> req)
             -> ss::future<ss::json::json_return_type> {
               auto auth_state = apply_auth<required_auth>(*req);
-
-              // Note: a request is only logged if it does not throw
-              // from authenticate().
-              log_request(*req, auth_state);
-
-              // Intercept exceptions
-              const auto url = req->get_url();
-              if constexpr (peek_auth) {
-                  return ss::futurize_invoke(
-                           handler, std::move(req), auth_state)
-                    .handle_exception(
-                      exception_intercepter<
-                        decltype(handler(std::move(req), auth_state).get0())>(
-                        url, auth_state));
-
-              } else {
-                  return ss::futurize_invoke(handler, std::move(req))
-                    .handle_exception(exception_intercepter<
-                                      decltype(handler(std::move(req)).get0())>(
-                      url, auth_state));
-              }
+              return ss::do_with(
+                std::move(auth_state),
+                [this, handler, req = std::move(req)](
+                  const auto& auth_state) mutable {
+                    // Note: a request is only logged if it does not throw
+                    // from authenticate().
+                    log_request(*req, auth_state);
```

Comment:
this solution is ok, but we should really be focusing on making coroutines not take reference parameters. case in point: the commit you referenced changed the suspension points, but then probably someone didn't go audit every caller.

here we could either move auth_state into the handlers after changing them to take values, or put it in a shared pointer and pass that to the handler and the exception handler continuation, or move into one of the two and copy into the other.

Reply:
Taking auth state by value is what I considered doing here. auth_state is small I don't think this is a perf critical path.